### PR TITLE
Common Calling Structure

### PR DESCRIPTION
**Pull request by ericlingerfelt**
## Description

This pull request allows all IODA converters written in python to be installable by pip and callable by module and package names. It also

- Adds the required `setup.py` file with all known python dependencies. 
- Adds the correct `__init__.py` files to the code.
- Adds the following code to `src/lib_python/orddicts.py` to manage the following deprecation so that python 3.10 can be used. See this [link](https://stackoverflow.com/questions/53978542/how-to-use-collections-abc-from-both-python-3-8-and-python-2-7).

```
try:
    from collections.abc import Callable
except ImportError:
    from collections import Callable
```
 
- Removes the following code from all files and includes the correct module imports to replace it. These classes can now be called with `lib_python` import statements. 

```
IODA_CONV_PATH = Path(__file__).parent/"@SCRIPT_LIB_PATH@"
if not IODA_CONV_PATH.is_dir():
    IODA_CONV_PATH = Path(__file__).parent/'..'/'lib-python'
sys.path.append(str(IODA_CONV_PATH.resolve()))
```

- Removes all unused imports. 
- Moves the script `generic_ob2ioda.py` to the top level in source. 
- Changes any files or directories in `src` to use `_` instead of `-` which is not allowed in python for modules or file names. 

Important notes: 

- This passes all ctests in `ioda-bundle` except `581 - test_iodaconv_airnow`. This is due to the converter's name being mismatched due to the change from `_` to `-`. The test file `test/testoutput/airnow_2020081306.nc` needs to have the top level attribute for converter updated to `airnow2ioda_nc.py` from `airnow2ioda-nc.py`.

### Issue(s) addressed

Link the issues to be closed with this PR
- fixes #<issue_number>

If this does not close an existing issue, then [be sure to add a Pipeline, Label, Estimate, Assignees, and Epic](https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/practices/issues.html)

## Acceptance Criteria (Definition of Done)

What does it mean for this PR to be finished?

## Dependencies

If there are PRs that need to be merged before or along with this one, please add "waiting for another PR" label and list the dependencies in the other repositories (example below). Note that the branches in the other repositories should have matching names for the automated tests to pass.
Waiting on the following PRs:
- [ ] waiting on JCSDA/eckit/pull/<pr_number>
- [ ] waiting on JCSDA/atlas/pull/<pr_number>

## Impact

Please list the other repositories (if any) that this PR will require changes in (example below)

- [ ] saber
- [ ] ufo
- [ ] ...

### CODE REVIEW

File: src/compo/airnow2ioda_nc.py [hunk 1]
```diff
@@ -2,19 +2,13 @@
...
-from pathlib import Path
-
-IODA_CONV_PATH = Path(__file__).parent/"@SCRIPT_LIB_PATH@"
-if not IODA_CONV_PATH.is_dir():
-    IODA_CONV_PATH = Path(__file__).parent/'..'/'lib-python'
-sys.path.append(str(IODA_CONV_PATH.resolve()))
-import meteo_utils
-import ioda_conv_ncio as iconv
+
+import lib_python.ioda_conv_engines as iconv
```
Comment (fabiolrdiniz) [inline src/compo/airnow2ioda_nc.py:8]:
```suggestion
```

**Review by fabiolrdiniz — APPROVED** [review]
Thanks!

**Comment by fabiolrdiniz** [comment]
I see this list of tests failing:
```
893 | 16 - test_iodaconv_gds2_sst_l2p (Failed)
894 | 17 - test_iodaconv_gds2_sst_l3u (Failed)
895 | 18 - test_iodaconv_smap_sss (Failed)
896 | 19 - test_iodaconv_rads_adt (Failed)
897 | 20 - test_iodaconv_godae_prof (Failed)
898 | 21 - test_iodaconv_godae_ship (Failed)
899 | 22 - test_iodaconv_godae_trak (Failed)
900 | 23 - test_iodaconv_hgodas_adt (Failed)
901 | 24 - test_iodaconv_hgodas_insitu (Failed)
902 | 25 - test_iodaconv_hgodas_sst (Failed)
903 | 26 - test_iodaconv_glider (Failed)
904 | 27 - test_iodaconv_hfradar (Failed)
905 | 28 - test_iodaconv_argoclim (Failed)
906 | 29 - test_iodaconv_cryosat2 (Failed)
907 | 30 - test_iodaconv_emc_ice (Failed)
908 | 31 - test_iodaconv_smos_sss (Failed)
909 | 32 - test_iodaconv_copernicus_l4adt (Failed)
910 | 33 - test_iodaconv_copernicus_l3swh (Failed)
911 | 34 - test_iodaconv_swot_l2adt (Failed)
912 | 35 - test_iodaconv_sst_ostia (Failed)
913 | 36 - test_iodaconv_avhrr_radiance (Failed)
914 | 37 - test_iodaconv_pace_oc_l2 (Failed)
915 | 38 - test_iodaconv_mls_o3_l2 (Failed)
916 | 39 - test_iodaconv_omi_o3_l2 (Failed)
917 | 40 - test_iodaconv_ompsnm_o3_l2 (Failed)
918 | 41 - test_iodaconv_satwinds (Failed)
919 | 42 - test_iodaconv_metar (Failed)
920 | 43 - test_iodaconv_viirs_jpss1_oc_l2 (Failed)
921 | 44 - test_iodaconv_modis_aqua_oc_l2 (Failed)
922 | 45 - test_iodaconv_viirs_jpss1_oc_l3 (Failed)
923 | 46 - test_iodaconv_godae_bgc_argo (Failed)
924 | 48 - test_iodaconv_gen_singleob (Failed)
925 | 49 - test_iodaconv_giirs_ssec2ioda (Failed)
926 | 50 - test_iodaconv_giirs_lw (Failed)
927 | 51 - test_iodaconv_gsidiag_conv_uv (Failed)
928 | 52 - test_iodaconv_gsidiag_conv (Failed)
929 | 53 - test_iodaconv_gsidiag_rad (Failed)
930 | 54 - test_iodaconv_wrfdadiag_rad (Failed)
931 | 56 - test_iodaconv_viirs_aod (Failed)
932 | 58 - test_iodaconv_tropomi_no2_total (Failed)
933 | 59 - test_iodaconv_tropomi_no2_tropo (Failed)
934 | 60 - test_iodaconv_mopitt_co (Failed)
935 | 61 - test_iodaconv_modis_aod (Failed)
936 | 62 - test_iodaconv_aeronet_aod (Failed)
937 | 63 - test_iodaconv_aeronet_aaod (Failed)
938 | 64 - test_iodaconv_airnow (Failed)
939 | 90 - test_iodaconv_generic_gnssro_bufr (Failed)
940 | 91 - test_iodaconv_atms_nc4 (Failed)
941 | 92 - test_iodaconv_generic_bufr_raob (Failed)
942 | 93 - test_iodaconv_amdar (Failed)
943 | 94 - test_iodaconv_buoy (Failed)
944 | 95 - test_iodaconv_ship (Failed)
945 | 96 - test_iodaconv_synop (Failed)
946 | 106 - test_iodaconv_ghcn_snod (Failed)
947 | 107 - test_iodaconv_smap_ssm (Failed)
948 | 108 - test_iodaconv_smos_ssm (Failed)
949 | 109 - test_iodaconv_ascat_ssm (Failed)
950 | 110 - test_iodaconv_imsfv3_scf (Failed)
951 | 111 - test_iodaconv_smap9km_ssm (Failed)
952 | 112 - test_iodaconv_land_owp_snow_obs_dup_thin (Failed)
```

**Comment by ericlingerfelt** [comment]
That's probably because of python dependencies not being installed. Have you ran `python3 -m pip install -e .` at the top level?

**Comment by fabiolrdiniz** [comment]
Sorry, @ericlingerfelt. I must have missed something. I meant in the CI tests. So probably the CI tests are not expected to pass at this point, right?

**Comment by ericlingerfelt** [comment]
> Sorry, @ericlingerfelt. I must have missed something. I meant in the CI tests. So probably the CI tests are not expected to pass at this point, right?

I don't know. I passed this work off to @BenjaminRuston.

**Comment by srherbener** [comment]
The CI tests should be passing. I'll take a quick look at the failures and let you know.

**Comment by srherbener** [comment]
Test are failing with this message:
```
Traceback (most recent call last):
  File "/build_container/bin/aeronet_aaod2ioda.py", line 28, in <module>
    import lib_python.ioda_conv_engines as iconv
ModuleNotFoundError: No module named 'lib_python'
```

I suspect that the CI configuration needs to be updated with the step that @ericlingerfelt mentioned:
```
Have you ran `python3 -m pip install -e .` at the top level?
```

**Comment by ericlingerfelt** [comment]
Yes. The main purpose of this PR is to make this an installable python package so that every python converter can be called from other python programs as a module.

**Comment by srherbener** [comment]
@mer-a-o can you help us get the test failures issue resolved? Is it easy to add the `python3 -m pip install -e .` in the CI configuration? 

@ericlingerfelt when should the pip install be run? It can be run at any time after the clone finishes (and the proper branch is checked out) and before running ctest. Is that correct? For example, would running the pip install be good right after ecbuild finishes?

**Comment by BenjaminRuston** [comment]
@fabiolrdiniz  there are a few,,, conflicts to resolve but this is exactly the directly we want to go!

**Comment by ericlingerfelt** [comment]
> @mer-a-o can you help us get the test failures issue resolved? Is it easy to add the `python3 -m pip install -e .` in the CI configuration?

When we add the testing harness to the other python repos (R2D2, for example), we will be using the approach we develop for `ioda-converters`. 

> 
> @ericlingerfelt when should the pip install be run? It can be run at any time after the clone finishes (and the proper branch is checked out) and before running ctest. Is that correct? For example, would running the pip install be good right after ecbuild finishes?

Hi @srherbener, yes, I think that is a correct statement. When I build and test `ioda-converters` locally, I have already installed the repo using `pip` and activated my `venv`.

**Comment by BenjaminRuston** [comment]
thanks everyone for looking at this I was going to work on some of there conflicts

**Comment by fabiolrdiniz** [comment]
> @srherbener Can you please list the new build steps for ioda-converter?

Not sure if it's the expected change for CI, but I did a test adding the `python3 -m pip install -e .` at some point after cloning & building this repo. Please, see https://github.com/JCSDA-internal/ioda-converters/pull/1087/commits/9ad6860019bef86e8dc3a765fa7c71d141638706.

**Comment by fabiolrdiniz** [comment]
Although CI tests passed (at least for CLang, as it's now), looks like the system is not so happy. See the warning below:
```bash
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
```

**Comment by mer-a-o** [comment]
Thanks @fabiolrdiniz I was going to make the same changes. I think we can ignore the warning about root user for CI.

**Comment by fabiolrdiniz** [comment]
Thanks, @mer-a-o! GNU is not happy:
```bash
929 | Installing collected packages: path, mccabe, pygrib, ioda-converters
930 | Running setup.py develop for ioda-converters
931 | error: subprocess-exited-with-error
932 |  
933 | × python setup.py develop did not run successfully.
934 | │ exit code: 1
935 | ╰─> [8 lines of output]
936 | running develop
937 | /opt/software/linux-ubuntu20.04-skylake_avx512/gcc-9.4.0/py-setuptools-59.4.0-sofyamgdqt4s3svq2bsbh4ywuaeckwhu/lib/python3.9/site-packages/setuptools/command/easy_install.py:156: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
938 | warnings.warn(
939 | /opt/software/linux-ubuntu20.04-skylake_avx512/gcc-9.4.0/py-setuptools-59.4.0-sofyamgdqt4s3svq2bsbh4ywuaeckwhu/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
940 | warnings.warn(
941 | running egg_info
942 | creating src/ioda_converters.egg-info
943 | error: could not create 'src/ioda_converters.egg-info': Permission denied
944 | [end of output]
945 |  
946 | note: This error originates from a subprocess, and is likely not a problem with pip.
947 | error: subprocess-exited-with-error
948 |  
949 | × python setup.py develop did not run successfully.
950 | │ exit code: 1
951 | ╰─> [8 lines of output]
952 | running develop
953 | /opt/software/linux-ubuntu20.04-skylake_avx512/gcc-9.4.0/py-setuptools-59.4.0-sofyamgdqt4s3svq2bsbh4ywuaeckwhu/lib/python3.9/site-packages/setuptools/command/easy_install.py:156: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
954 | warnings.warn(
955 | /opt/software/linux-ubuntu20.04-skylake_avx512/gcc-9.4.0/py-setuptools-59.4.0-sofyamgdqt4s3svq2bsbh4ywuaeckwhu/lib/python3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
956 | warnings.warn(
957 | running egg_info
958 | creating src/ioda_converters.egg-info
959 | error: could not create 'src/ioda_converters.egg-info': Permission denied
960 | [end of output]
961 |  
962 | note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Comment by fabiolrdiniz** [comment]
Thanks for the fix, @mer-a-o! All tests passed!

**Comment by BenjaminRuston** [comment]
@fcvdb this actually looks ready to roll!

this will be a huge upgrade can you have a look

File: src/gsi_ncdiag/proc_gsi_ncdiag.py [hunk 2]
```diff
@@ -6,11 +6,6 @@
 import time
 from pathlib import Path
 
-IODA_CONV_PATH = Path(__file__).parent/"@SCRIPT_LIB_PATH@"
-if not IODA_CONV_PATH.is_dir():
-    IODA_CONV_PATH = Path(__file__).parent/'..'/'lib-python'
-sys.path.append(str(IODA_CONV_PATH.resolve()))
-
 import gsi_ncdiag as gsid
```
Comment (CoryMartin-NOAA) [inline src/gsi_ncdiag/proc_gsi_ncdiag.py:9]: Does this path need to be updated? to `import gsi_ncdiag.gsi_ncdiag as gsid` like the next file is?

Comment (BenjaminRuston) [reply src/gsi_ncdiag/proc_gsi_ncdiag.py:9]: yes that's correct this will need to be changed

**Review by PatNichols — APPROVED** [review]

**Review by fcvdb — APPROVED** [review]
I built the branch and ran the tests. All tests passed, except those linked to `iodaconv_validate_testoutput, `but this is not related to the changes in the branch.

**Comment by PatNichols** [comment]
@ericlingerfelt can you resolve the conflicts above and fill in the PR synopsis above. Thanks!

**Comment by PatNichols** [comment]
@ericlingerfelt @BenjaminRuston  I took a stab at resolving the conflicts in this PR and develop. Eric you should definitely see if I broke something.

**Comment by ericlingerfelt** [comment]
HI All - I'm happy to but it may be a day or two before I find time. Let me know if this review is urgent and I can prioritize it.

**Comment by PatNichols** [comment]
@ericlingerfelt  yep I broke something

**Comment by ericlingerfelt** [comment]
> @ericlingerfelt yep I broke something

That's a lot of python. Easy to break something somewhere.

**Comment by BenjaminRuston** [comment]
@ericlingerfelt I can get to this later, so no high immediate priority but thanks to you both for getting this pushed forward!

**Comment by BenjaminRuston** [comment]
oh finally getting to reading this post,,, yes @ericlingerfelt in a day or two is perfect, looks like @PatNichols at least resolved the conflicts, we'll try to get this largely working and then have you take a look near the end to bless the changes

**Comment by ericlingerfelt** [comment]
Sounds good.

**Comment by BenjaminRuston** [comment]
@CoryMartin-NOAA , @ctgh and @DJDavies2 this went in!  Please let us know if you see any issue, was meaning hold off until you had a chance but please report any unexpected behavior or unintended consequences

**Comment by CoryMartin-NOAA** [comment]
How do we use this if we do not wish to 'pip install' but rather with $PYTHONPATH? Is that possible? Do we need to place a symlink to lib_python from pyiodaconv? Tagging @emilyhcliu as she noticed issues with using this in the way we used to

**Comment by climbfuji** [comment]
Can you do `python setup.py build` and not `python setup.py install` ? I haven't looked at the new functionality yet, but if that works you should be able to point PYTHONPATH to somewhere in your build directory structure?

**Comment by CoryMartin-NOAA** [comment]
what does ecbuild/cmake do as part of the bundle build process? The issue is that ecbuild seems to still put things in `pyiodaconv` and not `lib_python`, unless I'm missing something

**Comment by climbfuji** [comment]
> what does ecbuild/cmake do as part of the bundle build process? The issue is that ecbuild seems to still put things in `pyiodaconv` and not `lib_python`, unless I'm missing something

I haven't tried this myself, sorry ... let me know and I'll try to find some time to look into it

**Comment by CoryMartin-NOAA** [comment]
Sorry to be dumb here, but why are all of these now in the style of:
`import lib_python.ioda_conv_engines as iconv`

The name `lib_python` seems insanely generic to me. Am I just not understanding how this is supposed to be used? I would think it should be installed and used something like:
`import iodaconv.ioda_conv_engines as iconv`

**Comment by BenjaminRuston** [comment]
@CoryMartin-NOAA yes we should put some thought into this, 

generic yet specific could be `lib_jedi_python` or something along those lines too

there is an issue here of the not putting it into a common library. @PatNichols was going to start an issue to resolve that and can open one to name it appropriately...  thanks

**Comment by PatNichols** [comment]
Reverting this pull request

**Pull request merged**